numpy/core/src/multiarray/nditer_api.c

```diff
@@ -1390,6 +1390,41 @@ NpyIter_GetInnerLoopSizePtr(NpyIter *iter)
     }
 }
 
+/*NUMPY_API
+ * Resolves all writebackifcopy scratch buffers, not safe to use iterator
+ * operands after this call, in this iterator as well as any copies.
+ * Returns 0 on success, -1 on failure
+ */
+NPY_NO_EXPORT int
+NpyIter_Close(NpyIter *iter)
+{
+    int ret=0, iop, nop;
+    PyArrayObject ** operands;
+    npyiter_opitflags *op_itflags;
+    if (iter == NULL) {
+        return 0;
+    }
+    nop = NIT_NOP(iter);
+    operands = NIT_OPERANDS(iter);
+    op_itflags = NIT_OPITFLAGS(iter);
+    for (iop=0; iop<nop; iop++) {
+        if (op_itflags[iop] & NPY_OP_ITFLAG_HAS_WRITEBACK) {
+            op_itflags[iop] &= ~NPY_OP_ITFLAG_HAS_WRITEBACK;
+            if (PyArray_ResolveWritebackIfCopy(operands[iop]) < 0) {
+                ret = -1;
+                break;
+            }
+        }
+    }
+    for (; iop<nop; iop++) {
```

Comment:
Could do with a comment explaining what each of these two loops are.

Is this correct? The following happens right now:

```
PyArray_ResolveWritebackIfCopy(operands[0])
PyArray_ResolveWritebackIfCopy(operands[1])
PyArray_ResolveWritebackIfCopy(operands[2])  # FAIL
PyArray_DiscardWritebackIfCopy(operands[2])  # should this be repeated?
PyArray_DiscardWritebackIfCopy(operands[3])
PyArray_DiscardWritebackIfCopy(operands[4])

Reply:
good catch. It was harmless but redundant and wasteful. Fixed

Reply:
What state is the array left in if resolution fails? Perhaps it should be discarded after all?

Reply:
In short - no. A longer explanation - the operation that can fail in ``PyArray_ResolveWritebackIfCopy`` is a call to ``PyArray_CopyAnyInto``, all the other tasks are run to completion (resetting flags and base, DECREFing)

Reply:
Good to know, thanks. Is that / can that be documented?